dart.go

```diff
@@ -0,0 +1,756 @@
+	"false": {}, "final": {}, "finally": {}, "for": {}, "Function": {},
+	"get": {}, "hide": {}, "if": {}, "implements": {}, "import": {},
+	"in": {}, "interface": {}, "is": {}, "library": {}, "mixin": {},
+	"new": {}, "null": {}, "on": {}, "operator": {}, "part": {},
+	"rethrow": {}, "return": {}, "set": {}, "show": {}, "static": {},
+	"super": {}, "switch": {}, "sync": {}, "this": {}, "throw": {},
+	"true": {}, "try": {}, "typedef": {}, "var": {}, "void": {},
+	"while": {}, "with": {}, "yield": {},
+	// and some more not to interfere with properties
+	"a": {}, "buf": {}, "header": {}, "i": {}, "marshal": {}, "nextData": {},
+	"nextHeader": {}, "ns": {}, "readVarint": {}, "s": {}, "size": {},
+	"unmarshal": {}, "us": {}, "v": {}, "vi": {}, "view": {},
+}
+
+// GenerateDart writes the code into file "Colfer.dart".
+func GenerateDart(basedir string, packages Packages) error {
+	if err := os.MkdirAll(basedir, os.ModeDir|os.ModePerm); err != nil {
+		return err
+	}
+
+	t := template.New("dart-code")
+	template.Must(t.Parse(dartCode))
+	template.Must(t.New("marshal").Parse(dartMarshal))
+	template.Must(t.New("unmarshal").Parse(dartUnmarshal))
+
+	nativeTypes := map[string]string{
+		"text":      "String",
+		"binary":    "Uint8List",
+		"timestamp": "DateTime?",
+		"uint8":     "int",
+		"uint16":    "int",
+		"uint32":    "int",
+		"uint64":    "int",
+		"int32":     "int",
+		"int64":     "int",
+		"float32":   "double",
+		"float64":   "double",
+	}
+	nativeListTypes := map[string]string{
+		"uint8":   "Uint8List",
+		"uint16":  "Uint16List",
+		"uint32":  "Uint32List",
+		"uint64":  "Uint64List",
+		"int32":   "Int32List",
+		"int64":   "Int64List",
+		"float32": "Float32List",
+		"float64": "Float64List",
+	}
+
+	for _, p := range packages {
+		p.NameNative = p.Name
+		if _, ok := dartKeywords[p.NameNative]; ok {
+			p.NameNative += "_"
+		}
+		for _, t := range p.Structs {
+			t.NameNative = name.CamelCase(t.Name, true)
+			if _, ok := dartKeywords[t.NameNative]; ok {
+				t.NameNative += "_"
+			}
+			for _, f := range t.Fields {
+				f.NameNative = name.CamelCase(f.Name, false)
+				if _, ok := dartKeywords[f.NameNative]; ok {
+					f.NameNative += "_"
+				}
+			}
+		}
+	}
+
+	for _, p := range packages {
+		for _, t := range p.Structs {
+			for _, f := range t.Fields {
+				if f.TypeRef != nil {
+					f.TypeNative = f.TypeRef.NameNative
+					if f.TypeRef.Pkg != p {
+						f.TypeNative = f.TypeRef.Pkg.NameNative + "." + f.TypeNative
+					}
+					continue
+				}
+				if f.TypeList {
+					if nativeType, ok := nativeListTypes[f.Type]; ok {
+						f.TypeNative = nativeType
+					} else {
+						f.TypeNative = nativeTypes[f.Type]
+					}
+				} else {
+					if nativeType, ok := nativeTypes[f.Type]; ok {
+						f.TypeNative = nativeType
+					} else {
+						f.TypeNative = f.Type
+					}
+				}
+			}
+		}
+		f, err := os.Create(filepath.Join(basedir, fmt.Sprintf("colfer_%s.dart", p.Name)))
```

Comment:
Is the `_%s` suffix right here?

I assume there will/should be only one (generated) Colfer file per directory/package, right? Why not give it a fixed name, like 'Colfer.dart', just like: README, Makefile, .gitignore, etc.. The uppercase makes sure the least interesting files end up on top of an ls(1). Bit oldskool perhaps. 😆  Tell me what you think please.

Reply:
Dart likes relative imports, and I thought the easiest is to have it like this. If you run breaktest for Dart, it creates a directory named `breaktest` with two files in it: `colfer_static.dart` and `colfer_void.dart`. We could create a separate directory for them, and call both `Colfer.go` but for me `import '../static/Colfer.dart' as static_;` sounds worse than `import './colfer_static.dart' as static_;`, especially when we have multiple of these. But I can go with it if you prefer that.
https://dart-lang.github.io/linter/lints/prefer_relative_imports.html

Reply:
I believe having the package names in the directories is what the Dart site suggests for us to do, is it not?
Specifically, Colfer packages result into Dart packages, and Colfer types result into Dart classes (into the respective package). I.e., a Colfer package `foo` with type `A` and `B` goes into `./foo/Colfer.dart` containing class `A` and `B`. The compiler groups types by package and provides you with the base directory/path already.

https://github.com/pascaldekloe/colfer/blob/v1.8.1/go.go#L88

Reply:
Note that I'm not arguing against relative import paths. I just think we are supposed to use directories for packages, and map Colfer packages to Dart Packages.